## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [242]:
import pandas as pd

# Correct file path as a raw string
file_path = r"C:\Users\alexj\Documents\AI Bootcamp\Starter_Code\Starter_Code\Resources\client_dataset.csv"
df = pd.read_csv(file_path)

# Display the first few rows of the dataframe
print(df.head())


       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor  

In [243]:
# View the column names in the data

print(df.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [244]:
# Use the describe function to gather some basic statistics

df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [245]:
# Use this space to do any additional research
# and familiarize yourself with the data.


In [246]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [247]:
# For the category with the most entries, which subcategory had the most entries?
top_category = df['category'].value_counts().idxmax()
df[df['category'] == top_category]['subcategory'].value_counts().idxmax()

'bathroom supplies'

In [248]:
# Which five clients had the most entries in the data?
top_clients = df['client_id'].value_counts().head(5)
top_clients_list = top_clients.index.tolist()
print(top_clients)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [249]:
# Store the client ids of those top 5 clients in a list.

top_clients_list = top_clients.index.tolist()
print(top_clients_list)


[33615, 66037, 46820, 38378, 24741]


In [250]:
# How many total units (the qty column) did the client with the most entries order order?
top_client_id = top_clients_list[0]
total_units_top_client = df[df['client_id'] == top_client_id]['qty'].sum()
print(total_units_top_client)

64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [251]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']
df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [252]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.

df['shipping_price'] = df.apply(lambda row: 7 * row['unit_weight'] if row['unit_weight'] > 50 else 10 * row['unit_weight'], axis=1)

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  \
0     1096.80     762.71         7.50  105            1  115164.00   
1       24.95      15.09         1.49   21            0     523.95   
2       13.52       7.86         1.68   39            6     527.28   
3       36.42      24.85         1.23   29            3    1056.18   
4      195.10     108.17        46.43   20            1    3902.00   

   shipping_price  
0            75.0  
1            14.9  
2            16.8  
3            12.3  
4           464.3

In [253]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%

# Calculate sales tax
df['sales_tax'] = (df['subtotal'] + df['shipping_price']) * 0.0925
# Calculate total price including sales tax
df['total_price'] = df['subtotal'] + df['shipping_price'] + df['sales_tax']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number   subtotal  shipping_price     sales_tax  \
0         7.50  105            1  115164.00            75.0  10659.607500   
1         1.49   21            0     523.95            14.9     49.843625   
2         1.68   39            6     527.28            16.8     50.327400   
3         1.23   29            3    1056.18            12.3     98.834400   
4        46.43   20            1    3902.00           464.3    403.882750   

     total_price  
0  125898.607500  
1     588.693625  
2     594.407400  
3    1167.314400  
4    4770.182750  

[5 rows x 22 columns]

In [254]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).

df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']  
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   subtotal  shipping_price     sales_tax    total_price  \
0            1  115164.00            75.0  10659.607500  125898.607500   
1            0     523.95            14.9     49.843625     588.693625   
2            6     527.28            16.8     50.327400     594.407400   
3            3    1056.18            12.3     98.834400    1167.314400   
4            1    3902.00           464.3    403.882750    4770.182750   

   line_cost  
0   80159.55  
1     331.79  
2     323.34  
3     732.95  
4    2627.70  

[5 rows x 23 columns]

In [255]:
# Create a column for the profit of each line using line cost and line price

df['profit'] = df['total_price'] - df['line_cost']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

    subtotal  shipping_price     sales_tax    total_price  line_cost  \
0  115164.00            75.0  10659.607500  125898.607500   80159.55   
1     523.95            14.9     49.843625     588.693625     331.79   
2     527.28            16.8     50.327400     594.407400     323.34   
3    1056.18            12.3     98.834400    1167.314400     732.95   
4    3902.00           464.3    403.882750    4770.182750    2627.70   

         profit  
0  45739.057500  
1    256.903625  
2    271.067400  
3    434.364400  
4   2142.482750  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [256]:
# Check your work using the totals above

order_ids = [2742071, 2173913, 6128929]
for order_id in order_ids:
    order_lines = df[df['order_id'] == order_id]
    order_total = order_lines['total_price'].sum()
    print(f"Order ID {order_id} total price: ${order_total:.2f}")
    print(order_lines[['order_id', 'unit_price', 'qty', 'subtotal', 'unit_weight', 'shipping_price', 'sales_tax', 'total_price']])
    print("\n")



Order ID 2742071 total price: $139844.51
       order_id  unit_price  qty   subtotal  unit_weight  shipping_price  \
3446    2742071        5.59   77     430.43        10.46           104.6   
6287    2742071      460.20  260  119652.00         2.60            26.0   
20623   2742071      135.12   12    1621.44         2.27            22.7   
23840   2742071       47.59   22    1046.98         0.54             5.4   
24671   2742071       59.69   24    1432.56         1.57            15.7   
42470   2742071       82.18   44    3615.92         3.04            30.4   

          sales_tax    total_price  
3446      49.490275     584.520275  
6287   11070.215000  130748.215000  
20623    152.082950    1796.222950  
23840     97.345150    1149.725150  
24671    133.964050    1582.224050  
42470    337.284600    3983.604600  


Order ID 2173913 total price: $119838.36
       order_id  unit_price  qty  subtotal  unit_weight  shipping_price  \
627     2173913      110.27   70   7718.90       

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [257]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.

top_5_clients_revenue = df[df['client_id'].isin(top_clients_list)].groupby('client_id')['total_price'].sum()
top_5_clients_revenue

client_id
24741    7.667848e+07
33615    6.392918e+06
38378    9.172866e+06
46820    8.006333e+06
66037    8.746255e+06
Name: total_price, dtype: float64

In [258]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

summary_df = df[df['client_id'].isin(top_clients_list)].groupby('client_id').agg(
    total_units_purchased=('qty', 'sum'),
    total_shipping_price=('shipping_price', 'sum'),
    total_revenue=('total_price', 'sum'),
    total_profit=('profit', 'sum')
).reset_index()
summary_df

client_id  total_units_purchased  total_shipping_price  total_revenue  \
0      24741                 239862                9365.6   7.667848e+07   
1      33615                  64313               12609.4   6.392918e+06   
2      38378                  73667               11895.0   9.172866e+06   
3      46820                  75768               11094.8   8.006333e+06   
4      66037                  43018               10017.3   8.746255e+06   

   total_profit  
0  3.610666e+07  
1  2.033980e+06  
2  2.955706e+06  
3  2.589495e+06  
4  3.126907e+06

In [259]:
# Format the data and rename the columns to names suitable for presentation.
print("Columns in summary_df before formatting:", summary_df.columns)
# Define the money columns. 

money_columns = ['total_revenue', 'total_profit']

# Define a function that converts a dollar amount to millions.

def currency_format_millions(value):
    return value / 1_000_000


# Apply the currency_format_millions function to only the money columns. 

for col in money_columns:
    summary_df[col] = summary_df[col].apply(currency_format_millions)

# Rename the columns to reflect the change in the money format. 

summary_df = summary_df.rename(columns={
    'total_revenue': 'Total Revenue (in millions)',
    'total_profit': 'Total Profit (in millions)'
})
# Display the formatted DataFrame
summary_df.head()


Columns in summary_df before formatting: Index(['client_id', 'total_units_purchased', 'total_shipping_price',
       'total_revenue', 'total_profit'],
      dtype='object')


client_id  total_units_purchased  total_shipping_price  \
0      24741                 239862                9365.6   
1      33615                  64313               12609.4   
2      38378                  73667               11895.0   
3      46820                  75768               11094.8   
4      66037                  43018               10017.3   

   Total Revenue (in millions)  Total Profit (in millions)  
0                    76.678479                   36.106662  
1                     6.392918                    2.033980  
2                     9.172866                    2.955706  
3                     8.006333                    2.589495  
4                     8.746255                    3.126907

In [260]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.

# Function to convert values to millions of dollars
if 'total_revenue' in summary_df.columns and 'total_profit' in summary_df.columns:
    summary_df[['total_revenue', 'total_profit']] = summary_df[['total_revenue', 'total_profit']].applymap(to_millions)
summary_df


client_id  total_units_purchased  total_shipping_price  \
0      24741                 239862                9365.6   
1      33615                  64313               12609.4   
2      38378                  73667               11895.0   
3      46820                  75768               11094.8   
4      66037                  43018               10017.3   

   Total Revenue (in millions)  Total Profit (in millions)  
0                    76.678479                   36.106662  
1                     6.392918                    2.033980  
2                     9.172866                    2.955706  
3                     8.006333                    2.589495  
4                     8.746255                    3.126907

In [ ]:
## From looking at the data, I found that the top 5 clients are really
## important for the company’s revenue and profit. Client 24741 stands 
## out with the highest revenue and profit. This shows that keeping good
## relationships with these top clients is crucial for the business.
# Im having trouble with part 3 and the calculations.